### PR TITLE
[#1] Create Cpp stub simulation software

### PR DESCRIPTION
**Pull request by vectorlincent**
- Added ASIO as a Git submodule
- Added Cpp client and server code
- Added Kotlin client code

### CODE REVIEW

File: kotlin/src/Client.kt [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+fun main(args: Array<String>) {
+    try {
+        Client().use { client ->
+            val messageSent = "Hello C++!"
+            client.write(messageSent)
+            println("Sent message: \"$messageSent\"")
+            val messageReceived = client.read()
+            println("Received message: \"$messageReceived\"")
+            println("Handshake success.")
+        }
```
Comment (Groostav) [inline kotlin/src/Client.kt]: so @vectorlincent one of the problems with using a pipe in full-duplex like this is that it means we've got a race condition for who starts first: if both OASIS and Turbo-Tides are started at the same instance, who is going to create our server? can we get some mutex-gaurentees from the either the `java.net` or `asio` guys as a kind of server-lock?

Is it simpler to simply have our two processes create two statically-known servers, and have each communicate with the other in a simplex mode (which is how our existing plugin operates)?

File: .gitignore [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+hs_err_pid*
+
+# CLion project files
+*/.idea/workspace.xml
+
+# CMake output files
+cmake-build*/
+
+# Kotlin output files
+out/
```
Comment (Groostav) [inline .gitignore]: yeah this is idea's default, I dont like it much but it'l do for now.

File: .gitmodules [hunk 3]
```diff
@@ -0,0 +1,3 @@
+[submodule "cpp/extern/asio"]
+	path = cpp/extern/asio
+	url = https://github.com/chriskohlhoff/asio.git
```
Comment (Groostav) [inline .gitmodules]: hmm, is this standard-operating-procedure in C++ projects to add dependencies? Is there really no maven-like repo where we can just pull binaries?

File: kotlin/src/Client.kt [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+    override fun close() {
+        socket.close()
+        writer.close()
+        reader.close()
+    }
+
+    companion object {
+        val localHost = "127.0.0.1"
+        val localPort = 5000
```
Comment (Groostav) [inline kotlin/src/Client.kt]: Any reason you chose this number in particular?

Is it woth looking through [this long list](https://en.wikipedia.org/wiki/List_of_TCP_and_UDP_port_numbers) for a port that better fits us?

File: kotlin/src/Client.kt (hunk 1 above)
Comment (vectorlincent) [inline kotlin/src/Client.kt]: Hm sure. I was thinking one of them would just create a server end, and the other would create a client end.

File: kotlin/src/Client.kt (hunk 4 above)
Comment (vectorlincent) [inline kotlin/src/Client.kt]: I think 5000 was one of the port that Windows firewall recommended me to use when I tried to create a new rule in it.

File: .gitmodules (hunk 3 above)
Comment (vectorlincent) [inline .gitmodules]: Boost is largely template-based. So when you compile an application you'll likely recompile the header part of Boost anyway.

**Pull request merged**